src/view.c

```diff
@@ -54,6 +54,55 @@ view_edge_invert(enum view_edge edge)
 	}
 }
 
+struct edge_manip_hints {
+	int left_edge;
+	int top_edge;
+	int right_edge;
+	int bottom_edge;
+
+	int left_max_dx;
+	int up_max_dy;
+	int right_max_dx;
+	int down_max_dy;
+
+	int width_max_dx;
+	int height_max_dy;
+};
+
+static struct edge_manip_hints
+view_get_edge_manip_hints(struct view *view) {
+	struct output *output = view->output;
+	struct border margin = ssd_get_margin(view->ssd);
+	struct wlr_box usable = output_usable_area_in_layout_coords(output);
+	if (usable.height == output->wlr_output->height
+			&& output->wlr_output->scale != 1) {
+		usable.height /= output->wlr_output->scale;
+	}
+	if (usable.width == output->wlr_output->width
+			&& output->wlr_output->scale != 1) {
+		usable.width /= output->wlr_output->scale;
+	}
+
+	struct edge_manip_hints hints = {
+		.left_edge    = view->current.x - margin.left,
+		.top_edge     = view->current.y - margin.top,
+		.right_edge   = view->current.x + view->pending.width  + margin.right,
```

Comment:
I'm not 100% sure if I got the semantics of "pending" vs. "current" right. At what point are those values updated?

Reply:
Pending: whenever we change something
Current:
- when we apply the change ourselves (e.g. wayland native windows for a position only change)
- we get the confirmation of the window (e.g. wayland commit handler for new sizes).

Basically pending is always what the window "should" be and current is what labwc thinks it is right now (which may be outdated).

Reply:
This has moved to `snap_get_view_edge` and `snap_get_max_distance`:

https://github.com/digint/labwc/blob/441eb9031d9541beff60d740d6261c088dde0bd8/src/snap.c#L32

> So in this case for calculating the "to-be-set" state we should use ->pending for everything

Thanks for the explanation.

So you are saying I should change `view->current.x / y` as well? Will it not be confusing if we snap from something which is not "current == what labwc thinks it is right now" rather than some (virtual, to-be-changed in the future) window position? Does that mean that "current" is not always what I see on my screen?

I will gladly change it if you insist (as I'm still a bit puzzeled about this and don't fully understand how things are working under the hood)

Reply:
I'd say that in the unlikely case that there is indeed some pending movement / resize for a destination window (e.g. a window we want to snap to) that its nicer to see the new window go to the place where it should be and then having the target window adjust afterwards due to the pending commit than to position the window incorrectly once the pending commit comes through.